scio-core/src/main/scala/com/spotify/scio/io/BinaryIO.scala

```diff
@@ -42,14 +46,36 @@ import scala.util.Try
  *   a path to write to.
  */
 final case class BinaryIO(path: String) extends ScioIO[Array[Byte]] {
-  override type ReadP = Nothing
+  override type ReadP = BinaryIO.ReadParam
   override type WriteP = BinaryIO.WriteParam
+  // write options are insufficient to derive a reader for the output files
   override val tapT: TapT.Aux[Array[Byte], Nothing] = EmptyTapOf[Array[Byte]]
 
   override def testId: String = s"BinaryIO($path)"
 
-  override protected def read(sc: ScioContext, params: ReadP): SCollection[Array[Byte]] =
-    throw new UnsupportedOperationException("BinaryIO is write-only")
+  override protected def read(sc: ScioContext, params: ReadP): SCollection[Array[Byte]] = {
+    val filePattern = ScioUtil.filePattern(path, params.suffix)
+    val desiredBundleSizeBytes = 64 * 1024 * 1024L // 64 mb
+    val coder = ByteArrayCoder.of()
+    val srcFn = Functions.serializableFn { path: String =>
+      new BinaryIO.BinarySource(path, params.emptyMatchTreatment, params.reader)
+    }
+
+    sc.withName("Create filepattern")
+      .parallelize(List(filePattern))
+      .applyTransform("Match All", FileIO.matchAll())
+      .applyTransform(
+        "Read Matches",
```

Comment:
do we need to set transform names here? I'd think that Beam would set them under the hood for us.

Reply:
This replicates the internal usages in beam. There does not appear to be any naming applied from within the transform.